src/ai/backend/agent/agent.py

```diff
@@ -1476,6 +1479,24 @@ def _restore_ports(self, host_ports: Iterable[int]) -> None:
         ]
         self.port_pool.update(restored_ports)
 
+    def sync_port_pool(self, used_ports: Iterable[int]) -> None:
+        """
+        Synchronize the port pool with the actually used ports.
+        """
+        formatted_port_pool = ", ".join(map(str, self.port_pool))
+        formatted_used_ports = ", ".join(map(str, used_ports))
+        log.info(
+            "synchronizing port pool with used ports. current port pool: ({}), used ports: ({})",
+            formatted_port_pool,
+            formatted_used_ports,
+        )
+        port_range = self.local_config.container.port_range
+        used_port_set = set(used_ports)
+        original_port_pool: set[int] = {
+            p for p in range(port_range[0], port_range[1] + 1) if p not in used_port_set
+        }
+        self.port_pool = original_port_pool
```

Comment:
It seems like passing it all at once might cause timing issues. How about a method of removing ports that have been missed 'n' times or more?